File: typoswitch/layouts.py
"""QWERTY ↔ ЙЦУКЕН character maps (physical key positions)."""

from __future__ import annotations

EN_CHARS = "`qwertyuiop[]asdfghjkl;'zxcvbnm,./~QWERTYUIOP{}ASDFGHJKL:\"ZXCVBNM<>?"
RU_CHARS = "ёйцукенгшщзхъфывапролджэячсмитьбю.ЁЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЯЧСМИТЬБЮ,"

if len(EN_CHARS) != len(RU_CHARS):
    raise RuntimeError("Layout maps must be the same length")

EN_TO_RU = str.maketrans(EN_CHARS, RU_CHARS)
RU_TO_EN = str.maketrans(RU_CHARS, EN_CHARS)

LATIN = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`[];',./~{}:\"<>?")
CYRILLIC = set("абвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ")

WORD_CHARS = LATIN | CYRILLIC | set("-")
CONVERTIBLE = set(EN_CHARS) | set(RU_CHARS) | set("-")


def is_latin(text: str) -> bool:
    letters = [c for c in text if c.isalpha()]
    return bool(letters) and all(c in LATIN for c in letters)


def is_cyrillic(text: str) -> bool:
    letters = [c for c in text if c.isalpha()]
    return bool(letters) and all(c in CYRILLIC for c in letters)


def invert_layout(text: str) -> str:
    """Convert by opposite layout. Mixed script is converted char by char."""
    out = []
    for ch in text:
        if ch in LATIN:
            out.append(ch.translate(EN_TO_RU))
        elif ch in CYRILLIC:
            out.append(ch.translate(RU_TO_EN))
        else:
            out.append(ch)
    return "".join(out)


def majority_invert(text: str) -> str:
    """Invert the whole string toward the opposite of the dominant script."""
    latin = sum(1 for c in text if c in LATIN)
    cyr = sum(1 for c in text if c in CYRILLIC)
    if cyr > latin:
        return text.translate(RU_TO_EN)
    return text.translate(EN_TO_RU)

File: src/TypoSwitch.Core/Layouts.cs
namespace TypoSwitch;

public static class Layouts
{
    private const string EnChars = "`qwertyuiop[]asdfghjkl;'zxcvbnm,./~QWERTYUIOP{}ASDFGHJKL:\"ZXCVBNM<>?";
    private const string RuChars = "ёйцукенгшщзхъфывапролджэячсмитьбю.ЁЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЯЧСМИТЬБЮ,";

    private static readonly Dictionary<char, char> EnToRu = Build(EnChars, RuChars);
    private static readonly Dictionary<char, char> RuToEn = Build(RuChars, EnChars);

    public static readonly HashSet<char> Convertible = [..EnChars, ..RuChars, '-'];

    private static readonly HashSet<char> Latin = [.."`qwertyuiop[]asdfghjkl;'zxcvbnm,./~QWERTYUIOP{}ASDFGHJKL:\"ZXCVBNM<>?abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"];
    private static readonly HashSet<char> Cyrillic = [.. "абвгдеёжзийклмнопрстуфхцчшщъыьэюяАБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"];

    private static Dictionary<char, char> Build(string from, string to)
    {
        if (from.Length != to.Length)
            throw new InvalidOperationException("Layout maps must be the same length.");
        var map = new Dictionary<char, char>(from.Length);
        for (var i = 0; i < from.Length; i++)
            map[from[i]] = to[i];
        return map;
    }

    public static bool IsLatin(string text)
    {
        var any = false;
        foreach (var ch in text)
        {
            if (!char.IsLetter(ch)) continue;
            any = true;
            if (!Latin.Contains(ch)) return false;
        }
        return any;
    }

    public static bool IsCyrillic(string text)
    {
        var any = false;
        foreach (var ch in text)
        {
            if (!char.IsLetter(ch)) continue;
            any = true;
            if (!Cyrillic.Contains(ch)) return false;
        }
        return any;
    }

    public static string Invert(string text)
    {
        var chars = text.ToCharArray();
        for (var i = 0; i < chars.Length; i++)
        {
            var ch = chars[i];
            if (EnToRu.TryGetValue(ch, out var ru)) chars[i] = ru;
            else if (RuToEn.TryGetValue(ch, out var en)) chars[i] = en;
        }
        return new string(chars);
    }

    public static string MajorityInvert(string text)
    {
        var latin = 0;
        var cyr = 0;
        foreach (var ch in text)
        {
            if (Latin.Contains(ch)) latin++;
            else if (Cyrillic.Contains(ch)) cyr++;
        }
        var map = cyr > latin ? RuToEn : EnToRu;
        var chars = text.ToCharArray();
        for (var i = 0; i < chars.Length; i++)
        {
            if (map.TryGetValue(chars[i], out var mapped))
                chars[i] = mapped;
        }
        return new string(chars);
    }
}
